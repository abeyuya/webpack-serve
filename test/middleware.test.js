const { inspect } = require('util');

const webpack = require('webpack');

const ContentMiddleware = require('../lib/middleware/ContentMiddleware');
const MiddlewareState = require('../lib/middleware/MiddlewareState');
const WebpackMiddleware = require('../lib/middleware/WebpackMiddleware');

describe('middleware', () => {
  test('MiddlewareState', () => {
    const ware = new MiddlewareState();
    expect(ware.state).toBeDefined();
    expect(ware.called).toBe(false);
    ware.call();
    expect(ware.called).toBe(true);
  });

  test('MiddlewareState memo', () => {
    class TestMiddleware extends MiddlewareState {
      // eslint-disable-next-line class-methods-use-this
      call() {
        return +new Date();
      }
    }
    const ware = new TestMiddleware();
    expect(ware.state).toBeDefined();
    expect(ware.called).toBe(false);

    const result = ware.call();

    expect(ware.called).toBe(true);
    expect(ware.call()).toBe(result);
  });

  test('ContentMiddleware', () => {
    const app = {
      used: [],
      use(what) {
        app.used.push(inspect(what));
      },
    };
    const content = ['./batman', './superman'];
    const ware = new ContentMiddleware(app, { content });

    ware.call({});

    return ware.state.then(() => {
      expect(app.used).toMatchSnapshot();
    });
  });

  test('WebpackMiddleware', () => {
    const app = {
      used: [],
      use(what) {
        const value = inspect(what).replace(
          /startTime: \d+/g,
          'startTime: <START_TIME>'
        );
        app.used.push(value);
      },
    };
    const config = require('./fixtures/basic/webpack.config');
    const compiler = webpack(config);
    const logLevel = 'silent';
    const devMiddleware = { publicPath: '/', logLevel };
    const options = { compiler, devMiddleware, hotClient: { logLevel } };
    const ware = new WebpackMiddleware(app, options);

    return ware.call().then((koaWebpack) => {
      expect(app.used).toMatchSnapshot();
      expect(koaWebpack).toBeDefined();
      const value = inspect(koaWebpack).replace(/port: \d+/g, 'port: <PORT>');
      expect(value).toMatchSnapshot();

      const { hotClient } = koaWebpack;
      expect(hotClient.server.host).toBe('127.0.0.1');
      expect(hotClient.server.port).toBeGreaterThan(0);

      return new Promise((resolve) => koaWebpack.close(resolve));
    });
  });

  test('WebpackMiddleware Error', () => {
    const app = {
      use() {},
    };
    const config = require('./fixtures/basic/webpack.config');
    const compiler = webpack(config);
    const logLevel = 'silent';
    const devMiddleware = { publicPath: '/', logLevel };
    const options = {
      compiler,
      devMiddleware,
      hotClient: { logLevel, port: 'fail' },
    };
    const ware = new WebpackMiddleware(app, options);

    return Promise.all([
      ware.call().catch((error) => {
        expect(error).toMatchSnapshot();
      }),
      ware.state.then((result) => {
        expect(result instanceof Error).toBe(true);
        expect(result).toMatchSnapshot();
      }),
    ]);
  });
});
