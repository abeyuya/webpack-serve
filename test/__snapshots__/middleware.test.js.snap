// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware ContentMiddleware 1`] = `
Array [
  "[Function: serve]",
  "[Function: serve]",
]
`;

exports[`middleware WebpackMiddleware 1`] = `
Array [
  "hotClient",
  "devMiddleware",
  "close",
]
`;

exports[`middleware WebpackMiddleware Error 1`] = `
"webpack-hot-client

  Options Validation Error

  options.port  should be integer


  options.port  should be integer"
`;

exports[`middleware WebpackMiddleware Error 2`] = `
"Error: An error was thrown while initializing koa-webpack
 webpack-hot-client

  Options Validation Error

  options.port  should be integer


  options.port  should be integer"
`;
